{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pylab as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from time import time\n",
    "import logging\n",
    "import pickle\n",
    "from operator import itemgetter\n",
    "import json, pprint\n",
    "\n",
    "import argparse\n",
    "import yaml\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "\n",
    "from collections import defaultdict\n",
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "from typing import List, Dict, Union, Any\n",
    "\n",
    "from time import time\n",
    "\n",
    "SMALL_SIZE = 12\n",
    "MEDIUM_SIZE = 14\n",
    "BIGGER_SIZE = 16\n",
    "\n",
    "plt.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "plt.rc('axes', titlesize=SMALL_SIZE)     # fontsize of the axes title\n",
    "plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.INFO)\n",
    "log_info = logger.warning\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = argparse.ArgumentParser(description='Harmonize short-read and long-reads AnnData objects')\n",
    "\n",
    "parser.add_argument(\n",
    "    '-s',\n",
    "    '--short',\n",
    "    help='Path to harmonized short-reads AnnData objects',\n",
    "    required=True)\n",
    "\n",
    "parser.add_argument(\n",
    "    '-l',\n",
    "    '--long',\n",
    "    help='Path to harmonized long-reads AnnData objects',\n",
    "    required=True)\n",
    "\n",
    "parser.add_argument(\n",
    "    '-o',\n",
    "    '--output-path',\n",
    "    help='Output path',\n",
    "    required=True)\n",
    "\n",
    "parser.add_argument(\n",
    "    '-p',\n",
    "    '--output-prefix',\n",
    "    help='Output prefix',\n",
    "    required=True)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--min-cells-per-transcript',\n",
    "    default=0,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--min-cells-per-gene',\n",
    "    default=0,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--discovery-adata',\n",
    "    default='short',\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--barnyard-obs-column-name',\n",
    "    default='leiden_crude',\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--lo-expr-threshold',\n",
    "    default=1e-5,\n",
    "    type=float,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--hi-expr-threshold',\n",
    "    default=1e-4,\n",
    "    type=float,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--lo-expr-threshold-sensitive',\n",
    "    default=1e-6,\n",
    "    type=float,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--hi-expr-threshold-sensitive',\n",
    "    default=1e-4,\n",
    "    type=float,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--cell-purification-threshold',\n",
    "    default=1.,\n",
    "    type=float,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--contamination-threshold',\n",
    "    default=0.1,\n",
    "    type=float,\n",
    "    required=False)\n",
    "\n",
    "parser.add_argument(\n",
    "    '--min-umi-final',\n",
    "    default=100,\n",
    "    type=int,\n",
    "    required=False)\n",
    "\n",
    "# input\n",
    "notebook_mode = True\n",
    "\n",
    "if notebook_mode:\n",
    "    output_prefix = 'M132TS_final_10x_clr'\n",
    "    short_h5_path = f'/home/jupyter/mb-ml-dev-disk/MAS-seq-analysis/notebooks/cd8_t_cell_vdj_citeseq__paper__sandbox/barnyard_analysis/revised_v2/{output_prefix}.harmonized.barnyard.short.h5ad'\n",
    "    long_h5_path = f'/home/jupyter/mb-ml-dev-disk/MAS-seq-analysis/notebooks/cd8_t_cell_vdj_citeseq__paper__sandbox/barnyard_analysis/revised_v2/{output_prefix}.harmonized.barnyard.long.h5ad'\n",
    "    output_path = '/home/jupyter/mb-ml-dev-disk/MAS-seq-analysis/notebooks/cd8_t_cell_vdj_citeseq__paper__sandbox/barnyard_analysis/revised_v2'\n",
    "    args = f'-s {short_h5_path} -l {long_h5_path} -o {output_path} -p {output_prefix}'.split(' ')\n",
    "    args = vars(parser.parse_args(args))\n",
    "\n",
    "else:\n",
    "    args = vars(parser.parse_args())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input path\n",
    "final_long_adata_raw_h5_path = args['long']\n",
    "final_short_adata_raw_h5_path = args['short']\n",
    "output_path = args['output_path']\n",
    "output_prefix = args['output_prefix']\n",
    "\n",
    "# configuration\n",
    "min_cells_per_transcript = int(args['min_cells_per_transcript'])\n",
    "min_cells_per_gene = int(args['min_cells_per_gene'])\n",
    "group_cells_by_obs_key = args['barnyard_obs_column_name']\n",
    "\n",
    "# consts\n",
    "GENE_IDS_KEY = 'gene_ids'\n",
    "GENE_NAMES_KEY = 'gene_names'\n",
    "\n",
    "# group labels\n",
    "group_label_map = {\n",
    "    0: 'Immune',\n",
    "    1: 'Tumor'\n",
    "}\n",
    "\n",
    "# containers\n",
    "metrics = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('Loading long-reads AnnData object...')\n",
    "adata_long = sc.read(final_long_adata_raw_h5_path)\n",
    "assert group_cells_by_obs_key in adata_long.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('Filtering long-reads AnnData object...')\n",
    "# remove genes that are lowly expressed\n",
    "from collections import defaultdict\n",
    "gene_id_to_tx_indices_map = defaultdict(list)\n",
    "for i, gid in enumerate(adata_long.var[GENE_IDS_KEY].values):\n",
    "    gene_id_to_tx_indices_map[gid].append(i)\n",
    "\n",
    "included_gene_ids = []\n",
    "tx_counts_i = np.asarray(adata_long.X.sum(0)).flatten()\n",
    "for gid, tx_indices in gene_id_to_tx_indices_map.items():\n",
    "    if np.sum(tx_counts_i[tx_indices]) >= min_cells_per_gene:\n",
    "        included_gene_ids.append(gid)\n",
    "\n",
    "adata_long = adata_long[:, adata_long.var[GENE_IDS_KEY].values.isin(included_gene_ids)]\n",
    "\n",
    "# remove transcript that are very lowly expressed\n",
    "sc.pp.filter_genes(adata_long, min_cells=min_cells_per_transcript)\n",
    "total_umis = adata_long.X.sum()\n",
    "tpm_threshold = 1_000_000 * min_cells_per_transcript / total_umis\n",
    "\n",
    "log_info(f'Removing isoforms with TPM < {tpm_threshold:.2f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping from gene id to spanning tx icatces\n",
    "from collections import defaultdict\n",
    "gene_id_to_tx_indices_map = defaultdict(list)\n",
    "for i, gid in enumerate(adata_long.var[GENE_IDS_KEY].values):\n",
    "    gene_id_to_tx_indices_map[gid].append(i)\n",
    "\n",
    "# useful auxiliary data structures    \n",
    "gene_ids = sorted(list(gene_id_to_tx_indices_map.keys()))\n",
    "n_genes = len(gene_ids)\n",
    "n_transcripts = adata_long.shape[1]\n",
    "gene_id_to_gene_name_map = {\n",
    "    gene_id: gene_name for gene_id, gene_name in zip(adata_long.var[GENE_IDS_KEY], adata_long.var[GENE_NAMES_KEY])}\n",
    "gene_name_to_gene_id_map = {\n",
    "    gene_name: gene_id for gene_id, gene_name in zip(adata_long.var[GENE_IDS_KEY], adata_long.var[GENE_NAMES_KEY])}\n",
    "gene_names = list(map(gene_id_to_gene_name_map.get, gene_ids))\n",
    "\n",
    "# mapping from gene id to spanning tx indices\n",
    "group_ids = adata_long.obs[group_cells_by_obs_key].values.categories.values\n",
    "group_id_to_obs_indices_map = defaultdict(list)\n",
    "for group_id in group_ids:\n",
    "    group_id_to_obs_indices_map[group_id] = [\n",
    "        idx for idx in range(len(adata_long))\n",
    "        if adata_long.obs[group_cells_by_obs_key].values[idx] == group_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "\n",
    "# get gene expression from isoform expression\n",
    "row_indices = []\n",
    "col_indices = []\n",
    "values = []\n",
    "for j, gene_id in enumerate(gene_ids):\n",
    "    tx_indices = gene_id_to_tx_indices_map[gene_id]\n",
    "    row_indices += tx_indices\n",
    "    col_indices += [j] * len(tx_indices)\n",
    "    values += [1] * len(tx_indices)\n",
    "Y_ij = scipy.sparse.coo_matrix((values, (row_indices, col_indices)), shape=(n_transcripts, n_genes)).tocsr()\n",
    "gex_X_nj = adata_long.X @ Y_ij\n",
    "\n",
    "# normalize\n",
    "adata_long_gex = sc.AnnData(\n",
    "    X=gex_X_nj,\n",
    "    obs=adata_long.obs,\n",
    "    var=pd.DataFrame(index=pd.Index(list(map(gene_id_to_gene_name_map.get, gene_ids)))))\n",
    "\n",
    "adata_long_gex.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('Loading long-reads AnnData object...')\n",
    "adata_short = sc.read(final_short_adata_raw_h5_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_short_final = adata_short[:, adata_short.var.index.isin(adata_long_gex.var.index.values)]\n",
    "adata_long_final = adata_long_gex[:, adata_short_final.var.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "discovery_adata_name = args['discovery_adata']\n",
    "log_info(f'Identifying pure cells of each class from the discovery AnnData object ({discovery_adata_name})...')\n",
    "\n",
    "\n",
    "def get_grouped_expression(adata, group_cells_by_obs_key):\n",
    "    # mapping from gene id to spanning tx indices\n",
    "    group_ids = adata.obs[group_cells_by_obs_key].values.categories.values\n",
    "    group_id_to_obs_indices_map = defaultdict(list)\n",
    "    for group_id in group_ids:\n",
    "        group_id_to_obs_indices_map[group_id] = [\n",
    "            idx for idx in range(len(adata))\n",
    "            if adata.obs[group_cells_by_obs_key].values[idx] == group_id]\n",
    "    \n",
    "    n_genes = adata.shape[1]\n",
    "    n_groups = len(group_id_to_obs_indices_map)\n",
    "    group_expr_gi = np.zeros((n_groups, n_genes), dtype=np.int)\n",
    "    for i_group, group_id in enumerate(group_ids):\n",
    "        group_expr_gi[i_group, :] = np.asarray(adata.X[group_id_to_obs_indices_map[group_id], :].sum(0)).flatten()\n",
    "        \n",
    "    return group_expr_gi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('[First pass] identifying putative marker genes...')\n",
    "discovery_adata = {\n",
    "    'short': adata_short_final,\n",
    "    'long': adata_long_final}[discovery_adata_name]\n",
    "\n",
    "metacell_mg = get_grouped_expression(discovery_adata, group_cells_by_obs_key)\n",
    "normed_metacell_mg = metacell_mg / np.sum(metacell_mg, -1, keepdims=True)\n",
    "lo_expr_threshold = args['lo_expr_threshold']\n",
    "hi_expr_threshold = args['hi_expr_threshold']\n",
    "\n",
    "barnyard_gene_indices_list = []\n",
    "for group_a in range(2):\n",
    "    for group_b in range(2):\n",
    "        lo_in_a = normed_metacell_mg[group_a, :] < lo_expr_threshold\n",
    "        hi_in_b = normed_metacell_mg[group_b, :] > hi_expr_threshold\n",
    "        barnyard_mask_g = lo_in_a & hi_in_b\n",
    "        barnyard_gene_indices = np.where(barnyard_mask_g)[0]\n",
    "        for idx in barnyard_gene_indices:\n",
    "            barnyard_gene_indices_list.append((group_a, group_b, idx, discovery_adata.var.index.values[idx]))\n",
    "        metrics[f'n_genes_lo_group_{group_a}_hi_group_{group_b}__first_pass'] = int(barnyard_mask_g.sum()) \n",
    "        log_info(f'[First pass] Number of genes low in group {group_a}, high in group {group_b}: {barnyard_mask_g.sum()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "putative_group_1_gene_indices = [t[2] for t in barnyard_gene_indices_list if t[0] == 0 and t[1] == 1]\n",
    "putative_group_0_gene_indices = [t[2] for t in barnyard_gene_indices_list if t[0] == 1 and t[1] == 0]\n",
    "putative_group_1_gene_names = [t[3] for t in barnyard_gene_indices_list if t[0] == 0 and t[1] == 1]\n",
    "putative_group_0_gene_names = [t[3] for t in barnyard_gene_indices_list if t[0] == 1 and t[1] == 0]\n",
    "\n",
    "putative_group_1_gene_expr_in_group_0_n = np.asarray(\n",
    "    discovery_adata[discovery_adata.obs[group_cells_by_obs_key] == '0'][:, putative_group_1_gene_indices].X.sum(-1)).flatten()\n",
    "putative_group_0_gene_expr_in_group_1_n = np.asarray(\n",
    "    discovery_adata[discovery_adata.obs[group_cells_by_obs_key] == '1'][:, putative_group_0_gene_indices].X.sum(-1)).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('Saving putative marker genes diagnostics plot...')\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.hist(np.log1p(putative_group_0_gene_expr_in_group_1_n), bins=100);\n",
    "ax.set_xlabel('log1p(gex)')\n",
    "ax.set_ylabel('Number of cells')\n",
    "ax.set_title(f'Putative {group_label_map[0]} marker genes in {group_label_map[1]} cells')\n",
    "fig.tight_layout()\n",
    "plt.savefig(\n",
    "    os.path.join(\n",
    "        output_path,\n",
    "        f'{output_prefix}.barnyard.putative_01_gex.png'),\n",
    "    bbox_inches=\"tight\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.hist(np.log1p(putative_group_1_gene_expr_in_group_0_n), bins=100);\n",
    "ax.set_xlabel('log1p(gex)')\n",
    "ax.set_ylabel('Number of cells')\n",
    "ax.set_title(f'Putative {group_label_map[1]} marker genes in {group_label_map[0]} cells')\n",
    "fig.tight_layout()\n",
    "plt.savefig(\n",
    "    os.path.join(\n",
    "        output_path,\n",
    "        f'{output_prefix}.barnyard.putative_10_gex.png'),\n",
    "    bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('[First pass] Purifying cell types based on putative marker genes...')\n",
    "cell_purification_threshold = args['cell_purification_threshold']\n",
    "\n",
    "pure_group_0_cells_mask = putative_group_1_gene_expr_in_group_0_n < cell_purification_threshold\n",
    "pure_group_1_cells_mask = putative_group_0_gene_expr_in_group_1_n < cell_purification_threshold\n",
    "pure_group_0_cell_indices = np.where((discovery_adata.obs[group_cells_by_obs_key] == '0').values)[0][pure_group_0_cells_mask]\n",
    "pure_group_1_cell_indices = np.where((discovery_adata.obs[group_cells_by_obs_key] == '1').values)[0][pure_group_1_cells_mask]\n",
    "pure_both_indices = pure_group_0_cell_indices.tolist() + pure_group_1_cell_indices.tolist()\n",
    "\n",
    "metrics[f'n_0_cells__first_pass'] = int((discovery_adata.obs[group_cells_by_obs_key] == '0').values.sum())\n",
    "metrics[f'n_1_cells__first_pass'] = int((discovery_adata.obs[group_cells_by_obs_key] == '1').values.sum())\n",
    "metrics[f'n_0_cells__second_passs'] = len(pure_group_0_cell_indices)\n",
    "metrics[f'n_1_cells__second_pass'] = len(pure_group_1_cell_indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_short_final_pure = adata_short_final[pure_both_indices]\n",
    "adata_long_final_pure = adata_long_final[pure_both_indices]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('[Second pass] identifying highly specific marker genes...')\n",
    "discovery_adata = {\n",
    "    'short': adata_short_final_pure,\n",
    "    'long': adata_long_final_pure}[discovery_adata_name]\n",
    "\n",
    "metacell_mg = get_grouped_expression(adata_short_final_pure, group_cells_by_obs_key)\n",
    "normed_metacell_mg = metacell_mg / np.sum(metacell_mg, -1, keepdims=True)\n",
    "lo_expr_threshold = args['lo_expr_threshold_sensitive']\n",
    "hi_expr_threshold = args['hi_expr_threshold_sensitive']\n",
    "\n",
    "barnyard_gene_indices_list = []\n",
    "for group_a in range(2):\n",
    "    for group_b in range(2):\n",
    "        lo_in_a = normed_metacell_mg[group_a, :] < lo_expr_threshold\n",
    "        hi_in_b = normed_metacell_mg[group_b, :] > hi_expr_threshold\n",
    "        barnyard_mask_g = lo_in_a & hi_in_b\n",
    "        barnyard_gene_indices = np.where(barnyard_mask_g)[0]\n",
    "        for idx in barnyard_gene_indices:\n",
    "            barnyard_gene_indices_list.append((group_a, group_b, idx, discovery_adata.var.index.values[idx]))\n",
    "        metrics[f'n_genes_lo_group_{group_a}_hi_group_{group_b}__second_pass'] = int(barnyard_mask_g.sum())\n",
    "        log_info(f'[Second pass] Number of genes low in group {group_a}, high in group {group_b}: {barnyard_mask_g.sum()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_group_1_gene_indices = [t[2] for t in barnyard_gene_indices_list if t[0] == 0 and t[1] == 1]\n",
    "final_group_0_gene_indices = [t[2] for t in barnyard_gene_indices_list if t[0] == 1 and t[1] == 0]\n",
    "final_group_1_gene_names = [t[3] for t in barnyard_gene_indices_list if t[0] == 0 and t[1] == 1]\n",
    "final_group_0_gene_names = [t[3] for t in barnyard_gene_indices_list if t[0] == 1 and t[1] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5.5, 5))\n",
    "\n",
    "adata = adata_short_final_pure.copy()\n",
    "\n",
    "contamination_threshold = args['contamination_threshold']\n",
    "min_umi_final = args['min_umi_final']\n",
    "\n",
    "final_group_1_gene_expr_in_group_0_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '0'][:, final_group_1_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_0_gene_expr_in_group_1_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '1'][:, final_group_0_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_1_gene_expr_in_group_1_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '1'][:, final_group_1_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_0_gene_expr_in_group_0_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '0'][:, final_group_0_gene_indices].X.sum(-1)).flatten()\n",
    "\n",
    "outlier_group_1_cells_n = final_group_0_gene_expr_in_group_1_n > (contamination_threshold * final_group_1_gene_expr_in_group_1_n)\n",
    "outlier_group_1_cells_n = outlier_group_1_cells_n & (final_group_1_gene_expr_in_group_1_n > min_umi_final)\n",
    "outlier_group_0_cells_n = final_group_1_gene_expr_in_group_0_n > (contamination_threshold * final_group_0_gene_expr_in_group_0_n)\n",
    "outlier_group_0_cells_n = outlier_group_0_cells_n & (final_group_0_gene_expr_in_group_0_n > min_umi_final)\n",
    "\n",
    "# other statistics\n",
    "median_umi_per_cell = np.median(np.asarray(adata_short.X.sum(-1)).flat)\n",
    "group_1_purity = 100. * final_group_1_gene_expr_in_group_1_n.sum() / (final_group_0_gene_expr_in_group_1_n.sum() + final_group_1_gene_expr_in_group_1_n.sum())\n",
    "group_0_purity = 100. * final_group_0_gene_expr_in_group_0_n.sum() / (final_group_1_gene_expr_in_group_0_n.sum() + final_group_0_gene_expr_in_group_0_n.sum())\n",
    "n_group_0_cells = (adata.obs['leiden_crude'] == '0').sum()\n",
    "n_group_1_cells = (adata.obs['leiden_crude'] == '1').sum()\n",
    "\n",
    "metrics['n_outlier_group_0_cells__short'] = int(np.sum(outlier_group_0_cells_n))\n",
    "metrics['n_outlier_group_1_cells__short'] = int(np.sum(outlier_group_1_cells_n))\n",
    "metrics['median_umi_per_cell__short'] = int(median_umi_per_cell)\n",
    "metrics['group_0_purity__short'] = float(group_0_purity)\n",
    "metrics['group_1_purity__short'] = float(group_1_purity)\n",
    "metrics['n_group_0_cells__short'] = int(n_group_0_cells)\n",
    "metrics['n_group_1_cells__short'] = int(n_group_1_cells)\n",
    "\n",
    "# all points\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_0_n,\n",
    "    final_group_1_gene_expr_in_group_0_n,\n",
    "    s=1,\n",
    "    label=f'{group_label_map[0]} (N={n_group_0_cells})')\n",
    "\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_1_n,\n",
    "    final_group_1_gene_expr_in_group_1_n,\n",
    "    s=1,\n",
    "    label=f'{group_label_map[1]} (N={n_group_1_cells})')\n",
    "\n",
    "# outliers\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_0_n[outlier_group_0_cells_n],\n",
    "    final_group_1_gene_expr_in_group_0_n[outlier_group_0_cells_n],\n",
    "    s=50,\n",
    "    facecolor='none',\n",
    "    edgecolor='red',\n",
    "    marker='o',\n",
    "    lw=0.5,\n",
    "    label=f'{group_label_map[1]} in {group_label_map[0]} > {int(100. * contamination_threshold)}% (N={outlier_group_0_cells_n.sum()})')\n",
    "\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_1_n[outlier_group_1_cells_n],\n",
    "    final_group_1_gene_expr_in_group_1_n[outlier_group_1_cells_n],\n",
    "    s=50,\n",
    "    facecolor='none',\n",
    "    edgecolor='black',\n",
    "    marker='o',\n",
    "    lw=0.5,\n",
    "    label=f'{group_label_map[0]} in {group_label_map[1]} > {int(100. * contamination_threshold)}% (N={outlier_group_1_cells_n.sum()})')\n",
    "\n",
    "\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"Median UMIs per cell: {int(median_umi_per_cell)}\")\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"{group_label_map[1]} purity: {group_1_purity:.4f}%\")\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"{group_label_map[0]} purity: {group_0_purity:.4f}%\")\n",
    "\n",
    "ax.set_xlim((-40, 2000))\n",
    "ax.set_ylim((-40, 2000))\n",
    "\n",
    "ax.set_xlabel(f'{group_label_map[0]}-specific total GEX')\n",
    "ax.set_ylabel(f'{group_label_map[1]}-specific total GEX')\n",
    "\n",
    "# ax.set_title(f'Short-reads Dataset')\n",
    "ax.legend(fontsize=10)\n",
    "ax.grid(False)\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "log_info('Saving short-reads barnyard plot...')\n",
    "plt.savefig(\n",
    "    os.path.join(\n",
    "        output_path,\n",
    "        f'{output_prefix}.barnyard.short_reads.pdf'),\n",
    "    bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_0_cell_barcodes = adata_long_final_pure[adata_long_final_pure.obs['leiden_crude'] == '0'].obs.index.values\n",
    "group_1_cell_barcodes = adata_long_final_pure[adata_long_final_pure.obs['leiden_crude'] == '1'].obs.index.values\n",
    "group_0_fractional_library_share = adata_long_final[adata_long_final.obs.index.isin(group_0_cell_barcodes)].X.sum() / adata_long_final.X.sum()\n",
    "group_1_fractional_library_share = adata_long_final[adata_long_final.obs.index.isin(group_1_cell_barcodes)].X.sum() / adata_long_final.X.sum()\n",
    "group_0_total_expression = adata_long_final[:, final_group_0_gene_indices].X.sum()\n",
    "group_1_total_expression = adata_long_final[:, final_group_1_gene_indices].X.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "100 * (final_group_1_gene_expr_in_group_0_n.sum() / group_1_total_expression) / group_0_fractional_library_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "100 * (final_group_0_gene_expr_in_group_1_n.sum() / group_0_total_expression) / group_1_fractional_library_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5.5, 5))\n",
    "\n",
    "adata = adata_long_final_pure.copy()\n",
    "\n",
    "contamination_threshold = args['contamination_threshold']\n",
    "min_umi_final = args['min_umi_final']\n",
    "\n",
    "final_group_1_gene_expr_in_group_0_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '0'][:, final_group_1_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_0_gene_expr_in_group_1_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '1'][:, final_group_0_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_1_gene_expr_in_group_1_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '1'][:, final_group_1_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_0_gene_expr_in_group_0_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '0'][:, final_group_0_gene_indices].X.sum(-1)).flatten()\n",
    "\n",
    "outlier_group_1_cells_n = final_group_0_gene_expr_in_group_1_n > (contamination_threshold * final_group_1_gene_expr_in_group_1_n)\n",
    "outlier_group_1_cells_n = outlier_group_1_cells_n & (final_group_1_gene_expr_in_group_1_n > min_umi_final)\n",
    "outlier_group_0_cells_n = final_group_1_gene_expr_in_group_0_n > (contamination_threshold * final_group_0_gene_expr_in_group_0_n)\n",
    "outlier_group_0_cells_n = outlier_group_0_cells_n & (final_group_0_gene_expr_in_group_0_n > min_umi_final)\n",
    "\n",
    "# other statistics\n",
    "median_umi_per_cell = np.median(np.asarray(adata_long.X.sum(-1)).flat)\n",
    "group_1_purity = 100. * final_group_1_gene_expr_in_group_1_n.sum() / (final_group_0_gene_expr_in_group_1_n.sum() + final_group_1_gene_expr_in_group_1_n.sum())\n",
    "group_0_purity = 100. * final_group_0_gene_expr_in_group_0_n.sum() / (final_group_1_gene_expr_in_group_0_n.sum() + final_group_0_gene_expr_in_group_0_n.sum())\n",
    "n_group_0_cells = (adata.obs['leiden_crude'] == '0').sum()\n",
    "n_group_1_cells = (adata.obs['leiden_crude'] == '1').sum()\n",
    "\n",
    "metrics['n_outlier_group_0_cells__long'] = int(np.sum(outlier_group_0_cells_n))\n",
    "metrics['n_outlier_group_1_cells__long'] = int(np.sum(outlier_group_1_cells_n))\n",
    "metrics['median_umi_per_cell__long'] = int(median_umi_per_cell)\n",
    "metrics['group_0_purity__long'] = float(group_0_purity)\n",
    "metrics['group_1_purity__long'] = float(group_1_purity)\n",
    "metrics['n_group_0_cells__long'] = int(n_group_0_cells)\n",
    "metrics['n_group_1_cells__long'] = int(n_group_1_cells)\n",
    "\n",
    "# all points\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_0_n,\n",
    "    final_group_1_gene_expr_in_group_0_n,\n",
    "    s=1,\n",
    "    label=f'{group_label_map[0]} (N={n_group_0_cells})')\n",
    "\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_1_n,\n",
    "    final_group_1_gene_expr_in_group_1_n,\n",
    "    s=1,\n",
    "    label=f'{group_label_map[1]} (N={n_group_1_cells})')\n",
    "\n",
    "# outliers\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_0_n[outlier_group_0_cells_n],\n",
    "    final_group_1_gene_expr_in_group_0_n[outlier_group_0_cells_n],\n",
    "    s=50,\n",
    "    facecolor='none',\n",
    "    edgecolor='red',\n",
    "    marker='o',\n",
    "    lw=0.5,\n",
    "    label=f'{group_label_map[1]} in {group_label_map[0]} > {int(100. * contamination_threshold)}% (N={outlier_group_0_cells_n.sum()})')\n",
    "\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_1_n[outlier_group_1_cells_n],\n",
    "    final_group_1_gene_expr_in_group_1_n[outlier_group_1_cells_n],\n",
    "    s=50,\n",
    "    facecolor='none',\n",
    "    edgecolor='black',\n",
    "    marker='o',\n",
    "    lw=0.5,\n",
    "    label=f'{group_label_map[0]} in {group_label_map[1]} > {int(100. * contamination_threshold)}% (N={outlier_group_1_cells_n.sum()})')\n",
    "\n",
    "\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"Median UMIs per cell: {int(median_umi_per_cell)}\")\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"{group_label_map[1]} purity: {group_1_purity:.2f}%\")\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"{group_label_map[0]} purity: {group_0_purity:.2f}%\")\n",
    "\n",
    "# ax.set_xscale('log')\n",
    "# ax.set_yscale('log')\n",
    "\n",
    "\n",
    "ax.set_xlim((-20, 1000))\n",
    "ax.set_ylim((-20, 1000))\n",
    "\n",
    "ax.set_xlabel(f'{group_label_map[0]}-specific total GEX')\n",
    "ax.set_ylabel(f'{group_label_map[1]}-specific total GEX')\n",
    "\n",
    "# ax.set_title(f'Long-reads Dataset ({output_prefix})')\n",
    "ax.legend(fontsize=10)\n",
    "ax.grid(False)\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "log_info('Saving long-reads barnyard plot...')\n",
    "plt.savefig(\n",
    "    os.path.join(\n",
    "        output_path,\n",
    "        f'{output_prefix}.barnyard.long_reads.pdf'),\n",
    "    bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5.5, 5))\n",
    "\n",
    "adata = adata_long_final_pure.copy()\n",
    "\n",
    "contamination_threshold = args['contamination_threshold']\n",
    "min_umi_final = args['min_umi_final']\n",
    "\n",
    "final_group_1_gene_expr_in_group_0_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '0'][:, final_group_1_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_0_gene_expr_in_group_1_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '1'][:, final_group_0_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_1_gene_expr_in_group_1_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '1'][:, final_group_1_gene_indices].X.sum(-1)).flatten()\n",
    "final_group_0_gene_expr_in_group_0_n = np.asarray(\n",
    "    adata[adata.obs['leiden_crude'] == '0'][:, final_group_0_gene_indices].X.sum(-1)).flatten()\n",
    "\n",
    "outlier_group_1_cells_n = final_group_0_gene_expr_in_group_1_n > (contamination_threshold * final_group_1_gene_expr_in_group_1_n)\n",
    "outlier_group_1_cells_n = outlier_group_1_cells_n & (final_group_1_gene_expr_in_group_1_n > min_umi_final)\n",
    "outlier_group_0_cells_n = final_group_1_gene_expr_in_group_0_n > (contamination_threshold * final_group_0_gene_expr_in_group_0_n)\n",
    "outlier_group_0_cells_n = outlier_group_0_cells_n & (final_group_0_gene_expr_in_group_0_n > min_umi_final)\n",
    "\n",
    "# other statistics\n",
    "median_umi_per_cell = np.median(np.asarray(adata_long.X.sum(-1)).flat)\n",
    "group_1_purity = 100. * final_group_1_gene_expr_in_group_1_n.sum() / (final_group_0_gene_expr_in_group_1_n.sum() + final_group_1_gene_expr_in_group_1_n.sum())\n",
    "group_0_purity = 100. * final_group_0_gene_expr_in_group_0_n.sum() / (final_group_1_gene_expr_in_group_0_n.sum() + final_group_0_gene_expr_in_group_0_n.sum())\n",
    "n_group_0_cells = (adata.obs['leiden_crude'] == '0').sum()\n",
    "n_group_1_cells = (adata.obs['leiden_crude'] == '1').sum()\n",
    "\n",
    "metrics['n_outlier_group_0_cells__long'] = int(np.sum(outlier_group_0_cells_n))\n",
    "metrics['n_outlier_group_1_cells__long'] = int(np.sum(outlier_group_1_cells_n))\n",
    "metrics['median_umi_per_cell__long'] = int(median_umi_per_cell)\n",
    "metrics['group_0_purity__long'] = float(group_0_purity)\n",
    "metrics['group_1_purity__long'] = float(group_1_purity)\n",
    "metrics['n_group_0_cells__long'] = int(n_group_0_cells)\n",
    "metrics['n_group_1_cells__long'] = int(n_group_1_cells)\n",
    "\n",
    "# all points\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_0_n,\n",
    "    final_group_1_gene_expr_in_group_0_n,\n",
    "    s=4,\n",
    "    label=f'{group_label_map[0]}')\n",
    "\n",
    "ax.scatter(\n",
    "    final_group_0_gene_expr_in_group_1_n,\n",
    "    final_group_1_gene_expr_in_group_1_n,\n",
    "    s=4,\n",
    "    label=f'{group_label_map[1]}')\n",
    "\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"Median UMIs per cell: {int(median_umi_per_cell)}\")\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"{group_label_map[1]} purity: {group_1_purity:.2f}%\")\n",
    "plt.plot(\n",
    "    [], [], ' ',\n",
    "    label=f\"{group_label_map[0]} purity: {group_0_purity:.2f}%\")\n",
    "\n",
    "# ax.set_xscale('log')\n",
    "# ax.set_yscale('log')\n",
    "\n",
    "\n",
    "ax.set_xlim((-4, 200))\n",
    "ax.set_ylim((-4, 200))\n",
    "\n",
    "ax.set_xlabel(f'{group_label_map[0]}-specific total GEX', fontsize=20)\n",
    "ax.set_ylabel(f'{group_label_map[1]}-specific total GEX', fontsize=20)\n",
    "\n",
    "# ax.set_title(f'Long-reads Dataset ({output_prefix})')\n",
    "ax.legend(fontsize=14)\n",
    "ax.grid(False)\n",
    "ax.set_aspect('equal')\n",
    "plt.xticks(fontsize=16)\n",
    "plt.yticks(fontsize=16)\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "log_info('Saving long-reads barnyard plot...')\n",
    "plt.savefig(\n",
    "    os.path.join(\n",
    "        output_path,\n",
    "        f'{output_prefix}.barnyard.long_reads.pdf'),\n",
    "    bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "\n",
    "log_info('Saving metrics')\n",
    "with open(os.path.join(output_path, f'{output_prefix}.barnyard.metrics.yaml'), 'w+') as f:\n",
    "    yaml.dump(metrics, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_info('Done!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_0_cell_barcodes = adata_long_final_pure[adata_long_final_pure.obs['leiden_crude'] == '0'].obs.index.values\n",
    "group_1_cell_barcodes = adata_long_final_pure[adata_long_final_pure.obs['leiden_crude'] == '1'].obs.index.values\n",
    "\n",
    "group_0_fractional_library_share = adata_long_final[adata_long_final.obs.index.isin(group_0_cell_barcodes)].X.sum() / adata_long_final.X.sum()\n",
    "group_1_fractional_library_share = adata_long_final[adata_long_final.obs.index.isin(group_1_cell_barcodes)].X.sum() / adata_long_final.X.sum()\n",
    "\n",
    "group_0_total_expression = adata_long_final[:, final_group_0_gene_indices].X.sum()\n",
    "group_1_total_expression = adata_long_final[:, final_group_1_gene_indices].X.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "100 * (final_group_1_gene_expr_in_group_0_n.sum() / group_1_total_expression) / group_0_fractional_library_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "100 * (final_group_0_gene_expr_in_group_1_n.sum() / group_0_total_expression) / group_1_fractional_library_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "pytorch-gpu.1-7.m65",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/pytorch-gpu.1-7:m65"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
