{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential Expression (DE) and Differential Splicing (DS) analysis [M132TS Downsampling Analysis]\n",
    "\n",
    "**Inputs and Outputs**\n",
    "- Inputs:\n",
    "  - Harmonized and annotated short-read and long-read AnnData (raw, SCT)\n",
    "- Outputs:\n",
    "  - Figures\n",
    "  - Tables of global and per-cluster DE and DS pvalues for all genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %matplotlib inline\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import matplotlib.pylab as plt\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import logging\n",
    "from operator import itemgetter\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "from umap import UMAP\n",
    "\n",
    "SMALL_SIZE = 12\n",
    "MEDIUM_SIZE = 14\n",
    "BIGGER_SIZE = 16\n",
    "\n",
    "plt.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "plt.rc('axes', titlesize=SMALL_SIZE)     # fontsize of the axes title\n",
    "plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.INFO)\n",
    "log_info = logger.warning\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "mpl.rcParams.update(mpl.rcParamsDefault)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gencode_1m ...\n"
     ]
    }
   ],
   "source": [
    "# load sample metadata\n",
    "import yaml\n",
    "\n",
    "with open('./downsampling_series_sample_metadata.yaml', 'r') as f:\n",
    "    sample_meta_dict = yaml.safe_load(f)    \n",
    "\n",
    "notebook_mode = True\n",
    "\n",
    "if not notebook_mode:\n",
    "    sample_key = sys.argv[1]\n",
    "else:\n",
    "    sample_key = 'gencode_1m'\n",
    "    \n",
    "log_info(f'Processing {sample_key} ...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "repo_root = '/home/jupyter/mb-ml-data-disk/MAS-seq-analysis'\n",
    "long_tx_counts_root ='data/t-cell-vdj/long/quant/ds'\n",
    "short_h5_path = 'output/t-cell-vdj-cite-seq/M132TS_immune.h5ad'\n",
    "output_root = 'output/t-cell-vdj-cite-seq/ds'\n",
    "fig_output_root = 'output/t-cell-vdj-cite-seq/ds/figures'\n",
    "misc_output_root = 'output/t-cell-vdj-cite-seq/ds/misc'\n",
    "output_prefix = 'M132TS_immune'\n",
    "\n",
    "output_prefix_full = f'{output_prefix}_{sample_key}'\n",
    "\n",
    "final_long_adata_raw_h5_path = os.path.join(\n",
    "    repo_root, output_root, f'{output_prefix_full}_final_long_raw.h5ad')\n",
    "\n",
    "group_resolved_ds_df_csv_path = os.path.join(\n",
    "    repo_root, output_root, f'{output_prefix_full}_group_resolved_ds.csv')\n",
    "\n",
    "group_resolved_ds_de_df_csv_path =  os.path.join(\n",
    "    repo_root, output_root, f'{output_prefix_full}_group_resolved_ds_de.csv')\n",
    "\n",
    "global_ds_df_csv_path =  os.path.join(\n",
    "    repo_root, output_root, f'{output_prefix_full}_global_ds.csv')\n",
    "\n",
    "global_ds_de_df_csv_path =  os.path.join(\n",
    "    repo_root, output_root, f'{output_prefix_full}_global_ds_de.csv')\n",
    "\n",
    "# constants\n",
    "GENE_IDS_KEY = 'gencode_overlap_gene_ids'\n",
    "GENE_NAMES_KEY = 'gencode_overlap_gene_names'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_cells_per_transcript = 1\n",
    "min_cells_per_gene = 10\n",
    "\n",
    "group_cells_by_obs_key = 'mehrtash_leiden'\n",
    "\n",
    "n_mc_samples = 1_000_000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_long = sc.read(os.path.join(repo_root, final_long_adata_raw_h5_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5276 × 234552\n",
       "    obs: 'CD45_TotalSeqC', 'CD45R_B220_TotalSeqC', 'CD45RA_TotalSeqC', 'CD45RO_TotalSeqC', 'mehrtash_leiden'\n",
       "    var: 'transcript_ids', 'gene_ids', 'gene_names', 'transcript_names', 'de_novo_gene_ids', 'de_novo_transcript_ids', 'is_de_novo', 'is_gene_id_ambiguous', 'is_tcr_overlapping', 'gencode_overlap_gene_names', 'gencode_overlap_gene_ids', 'is_gencode_gene_overlap_ambiguous'\n",
       "    uns: 'mehrtash_leiden_colors'\n",
       "    obsm: 'X_pca_SCT_long', 'X_pca_SCT_short', 'X_pca_raw_short', 'X_tsne_raw_short', 'X_umap_SCT_long', 'X_umap_SCT_short'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Total UMIs: 329274.0\n"
     ]
    }
   ],
   "source": [
    "total_umis = adata_long.X.sum()\n",
    "log_info(f'Total UMIs: {total_umis}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.var` of view, copying.\n",
      "Removing isoforms with TPM < 3.04\n"
     ]
    }
   ],
   "source": [
    "# remove genes that are lowly expressed\n",
    "from collections import defaultdict\n",
    "gene_id_to_tx_indices_map = defaultdict(list)\n",
    "for i, gid in enumerate(adata_long.var[GENE_IDS_KEY].values):\n",
    "    gene_id_to_tx_indices_map[gid].append(i)\n",
    "\n",
    "included_gene_ids = []\n",
    "tx_counts_i = np.asarray(adata_long.X.sum(0)).flatten()\n",
    "for gid, tx_indices in gene_id_to_tx_indices_map.items():\n",
    "    if np.sum(tx_counts_i[tx_indices]) >= min_cells_per_gene:\n",
    "        included_gene_ids.append(gid)\n",
    "\n",
    "adata_long = adata_long[:, adata_long.var[GENE_IDS_KEY].values.isin(included_gene_ids)]\n",
    "\n",
    "# remove transcript that are very lowly expressed\n",
    "sc.pp.filter_genes(adata_long, min_cells=min_cells_per_transcript)\n",
    "tpm_threshold = 1_000_000 * min_cells_per_transcript / total_umis\n",
    "\n",
    "log_info(f'Removing isoforms with TPM < {tpm_threshold:.2f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5276 × 8807\n",
       "    obs: 'CD45_TotalSeqC', 'CD45R_B220_TotalSeqC', 'CD45RA_TotalSeqC', 'CD45RO_TotalSeqC', 'mehrtash_leiden'\n",
       "    var: 'transcript_ids', 'gene_ids', 'gene_names', 'transcript_names', 'de_novo_gene_ids', 'de_novo_transcript_ids', 'is_de_novo', 'is_gene_id_ambiguous', 'is_tcr_overlapping', 'gencode_overlap_gene_names', 'gencode_overlap_gene_ids', 'is_gencode_gene_overlap_ambiguous', 'n_cells'\n",
       "    uns: 'mehrtash_leiden_colors'\n",
       "    obsm: 'X_pca_SCT_long', 'X_pca_SCT_short', 'X_pca_raw_short', 'X_tsne_raw_short', 'X_umap_SCT_long', 'X_umap_SCT_short'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_long"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Isoform DE analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping from gene id to spanning tx icatces\n",
    "from collections import defaultdict\n",
    "gene_id_to_tx_indices_map = defaultdict(list)\n",
    "for i, gid in enumerate(adata_long.var[GENE_IDS_KEY].values):\n",
    "    gene_id_to_tx_indices_map[gid].append(i)\n",
    "\n",
    "# useful auxiliary data structures    \n",
    "gene_ids = sorted(list(gene_id_to_tx_indices_map.keys()))\n",
    "n_genes = len(gene_ids)\n",
    "gene_id_to_gene_name_map = {\n",
    "    gene_id: gene_name for gene_id, gene_name in zip(adata_long.var[GENE_IDS_KEY], adata_long.var[GENE_NAMES_KEY])}\n",
    "gene_name_to_gene_id_map = {\n",
    "    gene_name: gene_id for gene_id, gene_name in zip(adata_long.var[GENE_IDS_KEY], adata_long.var[GENE_NAMES_KEY])}\n",
    "gene_names = list(map(gene_id_to_gene_name_map.get, gene_ids))\n",
    "\n",
    "# mapping from gene id to spanning tx indices\n",
    "group_ids = adata_long.obs[group_cells_by_obs_key].values.categories.values\n",
    "group_id_to_obs_indices_map = defaultdict(list)\n",
    "for group_id in group_ids:\n",
    "    group_id_to_obs_indices_map[group_id] = [\n",
    "        idx for idx in range(len(adata_long))\n",
    "        if adata_long.obs[group_cells_by_obs_key].values[idx] == group_id]\n",
    "    \n",
    "# reduce tx expression by group (e.g. leiden clusters)\n",
    "n_transcripts = adata_long.shape[1]\n",
    "n_groups = len(group_id_to_obs_indices_map)\n",
    "group_expr_gi = np.zeros((n_groups, n_transcripts), dtype=np.int)\n",
    "for i_group, group_id in enumerate(group_ids):\n",
    "    group_expr_gi[i_group, :] = np.asarray(adata_long.X[group_id_to_obs_indices_map[group_id], :].sum(0)).flatten()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global differential splicing analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Dict, Tuple, Any, List\n",
    "from multiprocessing import Pool, cpu_count\n",
    "from tqdm import tqdm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from rpy2 import rinterface as ri\n",
    "from rpy2 import robjects\n",
    "from rpy2.rinterface_lib.embedded import RRuntimeError\n",
    "\n",
    "\n",
    "ri.initr()\n",
    "\n",
    "\n",
    "def get_global_ds_pval_fisher(\n",
    "        gene_id: str,\n",
    "        gene_id_to_tx_indices_map: Dict[str, List[int]],\n",
    "        group_expr_gi: np.ndarray,\n",
    "        n_mc_samples: int = 10_000) -> Dict[str, Any]:\n",
    "\n",
    "    tx_indices = gene_id_to_tx_indices_map[gene_id]\n",
    "    n_groups = group_expr_gi.shape[0]\n",
    "    x_gi = group_expr_gi[:, tx_indices]\n",
    "    x_g = np.sum(x_gi, -1)\n",
    "    n_nnz_groups = np.sum(x_g > 0)\n",
    "\n",
    "    if len(tx_indices) < 2 or n_nnz_groups < 2:\n",
    "        return {\n",
    "            'pval': 1.,\n",
    "            'x_g': x_g\n",
    "        }\n",
    "    \n",
    "    def rimport(packname):\n",
    "        as_environment = ri.baseenv['as.environment']\n",
    "        require = ri.baseenv['require']\n",
    "        require(ri.StrSexpVector([packname]),\n",
    "                quiet = ri.BoolSexpVector((True, )))\n",
    "        packname = ri.StrSexpVector(['package:' + str(packname)])\n",
    "        pack_env = as_environment(packname)\n",
    "        return pack_env\n",
    "    \n",
    "    rstats = rimport('stats')\n",
    "    \n",
    "    args = (('x', robjects.r.matrix(robjects.IntVector(x_gi.flat), nrow=n_groups, byrow=True)),\n",
    "            ('simulate.p.value', True),\n",
    "            ('B', n_mc_samples))\n",
    "    \n",
    "    try:\n",
    "        out = rstats['fisher.test'].rcall(args, ri.globalenv)\n",
    "        pval = float(np.array(np.array(out)[0])[0])\n",
    "        \n",
    "    except RRuntimeError:\n",
    "        log_info(f'Could not process {gene_id} -- returning p-value = 1. and continuing.')\n",
    "\n",
    "    return {\n",
    "        'pval': pval,\n",
    "        'x_g': x_g\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id_to_p_values_map = dict()\n",
    "gene_id_to_x_g_map = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2940/2940 [00:02<00:00, 1286.94it/s]\n"
     ]
    }
   ],
   "source": [
    "from typing import Tuple\n",
    "\n",
    "remaining_gene_ids = list(set(gene_ids).difference(gene_id_to_p_values_map.keys()))\n",
    "num_processes = cpu_count()\n",
    "\n",
    "def process_gene_id(gene_id: str) -> Dict[str, Any]:\n",
    "\n",
    "    out = get_global_ds_pval_fisher(\n",
    "        gene_id=gene_id,\n",
    "        gene_id_to_tx_indices_map=gene_id_to_tx_indices_map,\n",
    "        group_expr_gi=group_expr_gi,\n",
    "        n_mc_samples=n_mc_samples)\n",
    "    \n",
    "    return {\n",
    "        'gene_id': gene_id,\n",
    "        'pval': out['pval'],\n",
    "        'x_g': out['x_g']\n",
    "    }\n",
    "\n",
    "with Pool(processes=num_processes) as pool:\n",
    "    for result in tqdm(pool.imap(func=process_gene_id, iterable=remaining_gene_ids), total=len(remaining_gene_ids)):\n",
    "        gene_id_to_p_values_map[result['gene_id']] = result['pval']\n",
    "        gene_id_to_x_g_map[result['gene_id']] = result['x_g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate dataframe\n",
    "group_total_counts_dict = dict()\n",
    "for i_group, group_iad in enumerate(group_ids):\n",
    "    group_total_counts_dict[f'expr_{i_group}'] = list(map(lambda gene_id: gene_id_to_x_g_map[gene_id][i_group], gene_ids))\n",
    "group_total_counts_dict['total_expr'] = list(map(lambda gene_id: np.sum(gene_id_to_x_g_map[gene_id]), gene_ids))\n",
    "\n",
    "ds_pval_global = list(map(gene_id_to_p_values_map.get, gene_ids))\n",
    "_, ds_pval_global_adj, _, _ = multipletests(ds_pval_global, alpha=0.05, method='fdr_bh')\n",
    "\n",
    "global_ds_df = pd.DataFrame({\n",
    "    **dict(\n",
    "        gene_ids=gene_ids),\n",
    "    **group_total_counts_dict,\n",
    "    **dict(\n",
    "        ds_pval_global=ds_pval_global,\n",
    "        ds_pval_global_adj=ds_pval_global_adj)},\n",
    "    index=list(map(gene_id_to_gene_name_map.get, gene_ids)))\n",
    "\n",
    "global_ds_df.to_csv(global_ds_df_csv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>expr_0</th>\n",
       "      <th>expr_1</th>\n",
       "      <th>expr_2</th>\n",
       "      <th>expr_3</th>\n",
       "      <th>expr_4</th>\n",
       "      <th>expr_5</th>\n",
       "      <th>expr_6</th>\n",
       "      <th>expr_7</th>\n",
       "      <th>total_expr</th>\n",
       "      <th>ds_pval_global</th>\n",
       "      <th>ds_pval_global_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BAD</th>\n",
       "      <td>ENSG00000002330.14</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LAP3</th>\n",
       "      <td>ENSG00000002549.13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>0.867133</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD99</th>\n",
       "      <td>ENSG00000002586.20</td>\n",
       "      <td>26</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>35</td>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>9</td>\n",
       "      <td>146</td>\n",
       "      <td>0.632368</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD99_PAR_Y</th>\n",
       "      <td>ENSG00000002586.20_PAR_Y</td>\n",
       "      <td>29</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>4</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "      <td>45</td>\n",
       "      <td>14</td>\n",
       "      <td>162</td>\n",
       "      <td>0.711289</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAD1L1</th>\n",
       "      <td>ENSG00000002822.16</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>0.059940</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            gene_ids  expr_0  expr_1  expr_2  expr_3  expr_4  \\\n",
       "BAD               ENSG00000002330.14       2       0       2       0       1   \n",
       "LAP3              ENSG00000002549.13       0       2       4       0       5   \n",
       "CD99              ENSG00000002586.20      26       3      11       3      35   \n",
       "CD99_PAR_Y  ENSG00000002586.20_PAR_Y      29       6      18       4      28   \n",
       "MAD1L1            ENSG00000002822.16       3       2       0       0       3   \n",
       "\n",
       "            expr_5  expr_6  expr_7  total_expr  ds_pval_global  \\\n",
       "BAD              3       6       2          16        1.000000   \n",
       "LAP3             0       7       1          19        0.867133   \n",
       "CD99            19      40       9         146        0.632368   \n",
       "CD99_PAR_Y      18      45      14         162        0.711289   \n",
       "MAD1L1           1       3       6          18        0.059940   \n",
       "\n",
       "            ds_pval_global_adj  \n",
       "BAD                        1.0  \n",
       "LAP3                       1.0  \n",
       "CD99                       1.0  \n",
       "CD99_PAR_Y                 1.0  \n",
       "MAD1L1                     1.0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "global_ds_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Group-resolved differential splicing analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Dict, Tuple, Any\n",
    "from multiprocessing import Pool, cpu_count\n",
    "from tqdm import tqdm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from rpy2 import rinterface as ri\n",
    "from rpy2 import robjects\n",
    "from rpy2.rinterface_lib.embedded import RRuntimeError\n",
    "\n",
    "\n",
    "ri.initr()\n",
    "\n",
    "\n",
    "def get_group_resolved_ds_pval_fisher(\n",
    "        gene_id: str,\n",
    "        gene_id_to_tx_indices_map: Dict[str, List[int]],\n",
    "        group_expr_gi: np.ndarray,\n",
    "        n_mc_samples: int = 10_000) -> Dict[str, Any]:\n",
    "\n",
    "    tx_indices = gene_id_to_tx_indices_map[gene_id]\n",
    "    n_groups = group_expr_gi.shape[0]\n",
    "    x_gi = group_expr_gi[:, tx_indices]\n",
    "    x_g = np.sum(x_gi, -1)\n",
    "    \n",
    "    if len(tx_indices) < 2:\n",
    "        return {\n",
    "            'pval_g': np.ones((n_groups,)),\n",
    "            'x_g': x_g\n",
    "        }\n",
    "\n",
    "    # get leave-one-out sums\n",
    "    y_ggi = np.ascontiguousarray(np.repeat(x_gi[None, ...], n_groups, axis=0))\n",
    "    y_ggi[np.arange(n_groups), np.arange(n_groups), :] = 0.\n",
    "    y_gi = np.sum(y_ggi, axis=1)\n",
    "\n",
    "    def rimport(packname):\n",
    "        as_environment = ri.baseenv['as.environment']\n",
    "        require = ri.baseenv['require']\n",
    "        require(ri.StrSexpVector([packname]),\n",
    "                quiet = ri.BoolSexpVector((True, )))\n",
    "        packname = ri.StrSexpVector(['package:' + str(packname)])\n",
    "        pack_env = as_environment(packname)\n",
    "        return pack_env\n",
    "    \n",
    "    rstats = rimport('stats')\n",
    "\n",
    "    pval_g = np.zeros((n_groups,))\n",
    "    \n",
    "    for i_group in range(n_groups):\n",
    "        \n",
    "        x_i = x_gi[i_group, :]\n",
    "        y_i = y_gi[i_group, :]\n",
    "        \n",
    "        if np.sum(x_i) == 0 or np.sum(y_i) == 0:\n",
    "            pval = 1.\n",
    "        \n",
    "        else:            \n",
    "            contingency_table = np.hstack((x_i, y_i))\n",
    "            args = (('x', robjects.r.matrix(robjects.IntVector(contingency_table.flat), nrow=2, byrow=True)),\n",
    "                    ('simulate.p.value', True),\n",
    "                    ('B', n_mc_samples))\n",
    "            try:\n",
    "                out = rstats['fisher.test'].rcall(args, ri.globalenv)\n",
    "                pval = float(np.array(np.array(out)[0])[0])\n",
    "            except RRuntimeError:\n",
    "                print(contingency_table)\n",
    "                raise RuntimeError\n",
    "        \n",
    "        pval_g[i_group] = pval\n",
    "\n",
    "    return {\n",
    "        'pval_g': pval_g,\n",
    "        'x_g': x_g\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id_to_p_values_map = dict()\n",
    "gene_id_to_x_g_map = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2940/2940 [00:04<00:00, 672.97it/s]\n"
     ]
    }
   ],
   "source": [
    "from typing import Tuple\n",
    "\n",
    "remaining_gene_ids = list(set(gene_ids).difference(gene_id_to_p_values_map.keys()))\n",
    "num_processes = cpu_count()\n",
    "\n",
    "def process_gene_id(gene_id: str) -> Dict[str, Any]:\n",
    "\n",
    "    out = get_group_resolved_ds_pval_fisher(\n",
    "        gene_id=gene_id,\n",
    "        gene_id_to_tx_indices_map=gene_id_to_tx_indices_map,\n",
    "        group_expr_gi=group_expr_gi,\n",
    "        n_mc_samples=n_mc_samples)\n",
    "    \n",
    "    return {\n",
    "        'gene_id': gene_id,\n",
    "        'pval_g': out['pval_g'],\n",
    "        'x_g': out['x_g']\n",
    "    }\n",
    "\n",
    "with Pool(processes=num_processes) as pool:\n",
    "    for result in tqdm(pool.imap(func=process_gene_id, iterable=remaining_gene_ids), total=len(remaining_gene_ids)):\n",
    "        gene_id_to_p_values_map[result['gene_id']] = result['pval_g']\n",
    "        gene_id_to_x_g_map[result['gene_id']] = result['x_g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate dataframe\n",
    "pvalues_dict = dict()\n",
    "for i_group, group_id in enumerate(group_ids):\n",
    "    ds_pval_group = list(map(lambda gene_id: gene_id_to_p_values_map[gene_id][i_group], gene_ids))\n",
    "    _, ds_pval_group_adj, _, _ = multipletests(ds_pval_group, alpha=0.05, method='fdr_bh')\n",
    "    pvalues_dict[f'ds_pval_{i_group}'] = ds_pval_group\n",
    "    pvalues_dict[f'ds_pval_adj_{i_group}'] = ds_pval_group_adj\n",
    "\n",
    "group_total_counts_dict = dict()\n",
    "for i_group, group_id in enumerate(group_ids):\n",
    "    group_total_counts_dict[f'expr_{i_group}'] = list(map(lambda gene_id: gene_id_to_x_g_map[gene_id][i_group], gene_ids))\n",
    "group_total_counts_dict['total_expr'] = list(map(lambda gene_id: np.sum(gene_id_to_x_g_map[gene_id]), gene_ids))\n",
    "\n",
    "group_resolved_ds_df = pd.DataFrame({\n",
    "    **dict(gene_ids=gene_ids),\n",
    "    **pvalues_dict,\n",
    "    **group_total_counts_dict},\n",
    "    index=list(map(gene_id_to_gene_name_map.get, gene_ids)),)\n",
    "\n",
    "group_resolved_ds_df.to_csv(group_resolved_ds_df_csv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>ds_pval_0</th>\n",
       "      <th>ds_pval_adj_0</th>\n",
       "      <th>ds_pval_1</th>\n",
       "      <th>ds_pval_adj_1</th>\n",
       "      <th>ds_pval_2</th>\n",
       "      <th>ds_pval_adj_2</th>\n",
       "      <th>ds_pval_3</th>\n",
       "      <th>ds_pval_adj_3</th>\n",
       "      <th>ds_pval_4</th>\n",
       "      <th>...</th>\n",
       "      <th>ds_pval_adj_7</th>\n",
       "      <th>expr_0</th>\n",
       "      <th>expr_1</th>\n",
       "      <th>expr_2</th>\n",
       "      <th>expr_3</th>\n",
       "      <th>expr_4</th>\n",
       "      <th>expr_5</th>\n",
       "      <th>expr_6</th>\n",
       "      <th>expr_7</th>\n",
       "      <th>total_expr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BAD</th>\n",
       "      <td>ENSG00000002330.14</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LAP3</th>\n",
       "      <td>ENSG00000002549.13</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD99</th>\n",
       "      <td>ENSG00000002586.20</td>\n",
       "      <td>0.509491</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.529471</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.337662</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.409590</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>26</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>35</td>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>9</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD99_PAR_Y</th>\n",
       "      <td>ENSG00000002586.20_PAR_Y</td>\n",
       "      <td>0.780220</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.847153</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.432567</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>4</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "      <td>45</td>\n",
       "      <td>14</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAD1L1</th>\n",
       "      <td>ENSG00000002822.16</td>\n",
       "      <td>0.319680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.36963</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            gene_ids  ds_pval_0  ds_pval_adj_0  ds_pval_1  \\\n",
       "BAD               ENSG00000002330.14   1.000000            1.0    1.00000   \n",
       "LAP3              ENSG00000002549.13   1.000000            1.0    1.00000   \n",
       "CD99              ENSG00000002586.20   0.509491            1.0    1.00000   \n",
       "CD99_PAR_Y  ENSG00000002586.20_PAR_Y   0.780220            1.0    1.00000   \n",
       "MAD1L1            ENSG00000002822.16   0.319680            1.0    0.36963   \n",
       "\n",
       "            ds_pval_adj_1  ds_pval_2  ds_pval_adj_2  ds_pval_3  ds_pval_adj_3  \\\n",
       "BAD                   1.0   1.000000            1.0   1.000000            1.0   \n",
       "LAP3                  1.0   1.000000            1.0   1.000000            1.0   \n",
       "CD99                  1.0   0.529471            1.0   0.337662            1.0   \n",
       "CD99_PAR_Y            1.0   0.847153            1.0   0.432567            1.0   \n",
       "MAD1L1                1.0   1.000000            1.0   1.000000            1.0   \n",
       "\n",
       "            ds_pval_4  ...  ds_pval_adj_7  expr_0  expr_1  expr_2  expr_3  \\\n",
       "BAD          1.000000  ...            1.0       2       0       2       0   \n",
       "LAP3         1.000000  ...            1.0       0       2       4       0   \n",
       "CD99         0.409590  ...            1.0      26       3      11       3   \n",
       "CD99_PAR_Y   0.422577  ...            1.0      29       6      18       4   \n",
       "MAD1L1       0.714286  ...            1.0       3       2       0       0   \n",
       "\n",
       "            expr_4  expr_5  expr_6  expr_7  total_expr  \n",
       "BAD              1       3       6       2          16  \n",
       "LAP3             5       0       7       1          19  \n",
       "CD99            35      19      40       9         146  \n",
       "CD99_PAR_Y      28      18      45      14         162  \n",
       "MAD1L1           3       1       3       6          18  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_resolved_ds_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Group-resolved differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "\n",
    "# get gene expression from isoform expression\n",
    "row_indices = []\n",
    "col_indices = []\n",
    "values = []\n",
    "for j, gene_id in enumerate(gene_ids):\n",
    "    tx_indices = gene_id_to_tx_indices_map[gene_id]\n",
    "    row_indices += tx_indices\n",
    "    col_indices += [j] * len(tx_indices)\n",
    "    values += [1] * len(tx_indices)\n",
    "Y_ij = scipy.sparse.coo_matrix((values, (row_indices, col_indices)), shape=(n_transcripts, n_genes)).tocsr()\n",
    "gex_X_nj = adata_long.X @ Y_ij\n",
    "\n",
    "# normalize\n",
    "adata_long_gex = sc.AnnData(\n",
    "    X=gex_X_nj,\n",
    "    obs=adata_long.obs,\n",
    "    var=pd.DataFrame(index=pd.Index(list(map(gene_id_to_gene_name_map.get, gene_ids)))))\n",
    "\n",
    "adata_long_gex.var_names_make_unique()\n",
    "sc.pp.normalize_per_cell(adata_long_gex)\n",
    "sc.pp.log1p(adata_long_gex)\n",
    "\n",
    "adata_short_sct = adata_long_gex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_resolved_ds_df = pd.read_csv(group_resolved_ds_df_csv_path, index_col=0)\n",
    "\n",
    "sc.tl.rank_genes_groups(adata_short_sct, 'mehrtash_leiden', method='t-test', use_raw=False)\n",
    "\n",
    "result = adata_short_sct.uns['rank_genes_groups']\n",
    "n_groups = len(group_ids)\n",
    "group_id_to_group_idx_map = {group_id: group_idx for group_idx, group_id in enumerate(group_ids)}\n",
    "key_to_col_name_map = {'names': 'gene_names', 'pvals': 'de_pval', 'pvals_adj': 'de_pval_adj_scanpy'}\n",
    "de_df = pd.DataFrame(\n",
    "    {key_to_col_name_map[key] + '_' + str(group_id_to_group_idx_map[group_id]): result[key][group_id]\n",
    "     for group_id in group_ids\n",
    "     for key in ['names', 'pvals', 'pvals_adj']})\n",
    "\n",
    "de_gene_names_list = adata_short_sct.var.index.values\n",
    "de_pval_list = np.zeros((len(de_gene_names_list),))\n",
    "de_gene_name_to_index_map = {gene_name: idx for idx, gene_name in enumerate(de_gene_names_list)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "ds_gene_names_list = group_resolved_ds_df.index.values.tolist()\n",
    "mutual_gene_names_list = list(set(ds_gene_names_list).intersection(de_gene_names_list))\n",
    "group_resolved_ds_df = group_resolved_ds_df.loc[mutual_gene_names_list]\n",
    "\n",
    "gene_name_counter = Counter(group_resolved_ds_df.index.values)\n",
    "bad_gene_names = set()\n",
    "for gene_name, multiplicity in gene_name_counter.items():\n",
    "    if multiplicity > 1:\n",
    "        bad_gene_names.add(gene_name)\n",
    "\n",
    "mutual_gene_names_list = list(set(mutual_gene_names_list).difference(bad_gene_names))\n",
    "group_resolved_ds_df = group_resolved_ds_df.loc[mutual_gene_names_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_resolved_ds_de_df = group_resolved_ds_df.copy()\n",
    "\n",
    "for group_idx, group_id in enumerate(group_ids):\n",
    "    de_group_gene_names = de_df[f'gene_names_{group_idx}'].values\n",
    "    de_group_pvals = de_df[f'de_pval_{group_idx}'].values\n",
    "    _, de_group_pvals_adj, _, _ = multipletests(de_group_pvals, alpha=0.05, method='fdr_bh')\n",
    "    de_group_gene_name_to_index_map = {gene_name: index for index, gene_name in enumerate(de_group_gene_names)}\n",
    "    group_resolved_ds_de_df[f'de_pval_{group_idx}'] = de_group_pvals[\n",
    "        list(map(de_group_gene_name_to_index_map.get, mutual_gene_names_list))]\n",
    "    group_resolved_ds_de_df[f'de_pval_adj_{group_idx}'] = de_group_pvals_adj[\n",
    "        list(map(de_group_gene_name_to_index_map.get, mutual_gene_names_list))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_resolved_ds_de_df.to_csv(group_resolved_ds_de_df_csv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>ds_pval_0</th>\n",
       "      <th>ds_pval_adj_0</th>\n",
       "      <th>ds_pval_1</th>\n",
       "      <th>ds_pval_adj_1</th>\n",
       "      <th>ds_pval_2</th>\n",
       "      <th>ds_pval_adj_2</th>\n",
       "      <th>ds_pval_3</th>\n",
       "      <th>ds_pval_adj_3</th>\n",
       "      <th>ds_pval_4</th>\n",
       "      <th>...</th>\n",
       "      <th>de_pval_3</th>\n",
       "      <th>de_pval_adj_3</th>\n",
       "      <th>de_pval_4</th>\n",
       "      <th>de_pval_adj_4</th>\n",
       "      <th>de_pval_5</th>\n",
       "      <th>de_pval_adj_5</th>\n",
       "      <th>de_pval_6</th>\n",
       "      <th>de_pval_adj_6</th>\n",
       "      <th>de_pval_7</th>\n",
       "      <th>de_pval_adj_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IFNAR2</th>\n",
       "      <td>ENSG00000159110.20</td>\n",
       "      <td>0.327672</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.427572</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.088051</td>\n",
       "      <td>0.350878</td>\n",
       "      <td>0.104540</td>\n",
       "      <td>0.365455</td>\n",
       "      <td>0.270833</td>\n",
       "      <td>0.496725</td>\n",
       "      <td>9.795041e-01</td>\n",
       "      <td>9.872472e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL12B</th>\n",
       "      <td>ENSG00000118680.14</td>\n",
       "      <td>0.243756</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.398601</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.015984</td>\n",
       "      <td>...</td>\n",
       "      <td>0.515332</td>\n",
       "      <td>0.578494</td>\n",
       "      <td>0.005880</td>\n",
       "      <td>0.054024</td>\n",
       "      <td>0.812816</td>\n",
       "      <td>0.959380</td>\n",
       "      <td>0.001316</td>\n",
       "      <td>0.014228</td>\n",
       "      <td>1.714555e-11</td>\n",
       "      <td>6.223198e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC38A2</th>\n",
       "      <td>ENSG00000134294.14</td>\n",
       "      <td>0.154799</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.395500</td>\n",
       "      <td>0.481080</td>\n",
       "      <td>0.248958</td>\n",
       "      <td>0.584614</td>\n",
       "      <td>0.206025</td>\n",
       "      <td>0.568527</td>\n",
       "      <td>0.890368</td>\n",
       "      <td>0.945695</td>\n",
       "      <td>8.103127e-01</td>\n",
       "      <td>8.981456e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP5F1A</th>\n",
       "      <td>ENSG00000152234.16</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.748252</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.202797</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.436563</td>\n",
       "      <td>...</td>\n",
       "      <td>0.565134</td>\n",
       "      <td>0.619268</td>\n",
       "      <td>0.833579</td>\n",
       "      <td>0.925224</td>\n",
       "      <td>0.071002</td>\n",
       "      <td>0.281328</td>\n",
       "      <td>0.485082</td>\n",
       "      <td>0.676159</td>\n",
       "      <td>1.256513e-01</td>\n",
       "      <td>2.854828e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDF2</th>\n",
       "      <td>ENSG00000132581.10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002861</td>\n",
       "      <td>0.004574</td>\n",
       "      <td>0.002859</td>\n",
       "      <td>0.031367</td>\n",
       "      <td>0.897391</td>\n",
       "      <td>0.979022</td>\n",
       "      <td>0.420090</td>\n",
       "      <td>0.629129</td>\n",
       "      <td>6.710085e-01</td>\n",
       "      <td>8.115035e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_ids  ds_pval_0  ds_pval_adj_0  ds_pval_1  \\\n",
       "IFNAR2   ENSG00000159110.20   0.327672            1.0   1.000000   \n",
       "MYL12B   ENSG00000118680.14   0.243756            1.0   0.398601   \n",
       "SLC38A2  ENSG00000134294.14   0.154799            1.0   1.000000   \n",
       "ATP5F1A  ENSG00000152234.16   1.000000            1.0   1.000000   \n",
       "SDF2     ENSG00000132581.10   1.000000            1.0   1.000000   \n",
       "\n",
       "         ds_pval_adj_1  ds_pval_2  ds_pval_adj_2  ds_pval_3  ds_pval_adj_3  \\\n",
       "IFNAR2             1.0   1.000000            1.0   1.000000            1.0   \n",
       "MYL12B             1.0   1.000000            1.0   1.000000            1.0   \n",
       "SLC38A2            1.0   1.000000            1.0   0.263158            1.0   \n",
       "ATP5F1A            1.0   0.748252            1.0   0.202797            1.0   \n",
       "SDF2               1.0   1.000000            1.0   1.000000            1.0   \n",
       "\n",
       "         ds_pval_4  ...  de_pval_3  de_pval_adj_3  de_pval_4  de_pval_adj_4  \\\n",
       "IFNAR2    0.427572  ...   0.000052       0.000160   0.088051       0.350878   \n",
       "MYL12B    0.015984  ...   0.515332       0.578494   0.005880       0.054024   \n",
       "SLC38A2   1.000000  ...   0.395500       0.481080   0.248958       0.584614   \n",
       "ATP5F1A   0.436563  ...   0.565134       0.619268   0.833579       0.925224   \n",
       "SDF2      1.000000  ...   0.002861       0.004574   0.002859       0.031367   \n",
       "\n",
       "         de_pval_5  de_pval_adj_5  de_pval_6  de_pval_adj_6     de_pval_7  \\\n",
       "IFNAR2    0.104540       0.365455   0.270833       0.496725  9.795041e-01   \n",
       "MYL12B    0.812816       0.959380   0.001316       0.014228  1.714555e-11   \n",
       "SLC38A2   0.206025       0.568527   0.890368       0.945695  8.103127e-01   \n",
       "ATP5F1A   0.071002       0.281328   0.485082       0.676159  1.256513e-01   \n",
       "SDF2      0.897391       0.979022   0.420090       0.629129  6.710085e-01   \n",
       "\n",
       "         de_pval_adj_7  \n",
       "IFNAR2    9.872472e-01  \n",
       "MYL12B    6.223198e-10  \n",
       "SLC38A2   8.981456e-01  \n",
       "ATP5F1A   2.854828e-01  \n",
       "SDF2      8.115035e-01  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_resolved_ds_de_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "ncols = 4\n",
    "nrows = int(np.ceil(len(group_ids) / ncols))\n",
    "scale = 3.\n",
    "\n",
    "fig, axs = plt.subplots(ncols=ncols, nrows=nrows, figsize=(scale * ncols, scale * nrows))\n",
    "\n",
    "highlight_map = {\n",
    "    # ('RPL', 'red'): [gene_name for gene_name in gene_names if gene_name.find('RPL') == 0],\n",
    "    ('PTPRC', 'blue'): ['PTPRC'],\n",
    "}\n",
    "\n",
    "x_jitter_scale = 0.01\n",
    "y_jitter_scale = 0.01\n",
    "max_ds_log_pval = 3.\n",
    "max_de_log_pval = 3.\n",
    "log10_fdr_threshold = - np.log10(0.05)\n",
    "rng_seed = 0\n",
    "\n",
    "rng = np.random.RandomState(seed=rng_seed)\n",
    "gene_names_list = group_resolved_ds_de_df.index.values.tolist()\n",
    "log10_tpm = np.log10(\n",
    "    1_000_000 * group_resolved_ds_de_df['total_expr'].values /\n",
    "    np.sum(group_resolved_ds_de_df['total_expr'].values))\n",
    "\n",
    "for group_idx, ax in zip(range(len(group_ids)), axs.flatten()):\n",
    "\n",
    "    X_COL_NAME = f'de_pval_adj_{group_idx}'\n",
    "    Y_COL_NAME = f'ds_pval_adj_{group_idx}'\n",
    "\n",
    "    xx = np.minimum(-np.log10(group_resolved_ds_de_df[X_COL_NAME].values), max_de_log_pval)\n",
    "    yy = np.minimum(-np.log10(group_resolved_ds_de_df[Y_COL_NAME].values), max_ds_log_pval)\n",
    "\n",
    "    n_ds = np.sum((yy > log10_fdr_threshold) & (xx <= log10_fdr_threshold))\n",
    "    n_de = np.sum((xx > log10_fdr_threshold) & (yy <= log10_fdr_threshold))\n",
    "    n_de_ds = np.sum((xx > log10_fdr_threshold) & (yy > log10_fdr_threshold))\n",
    "    n_boring = np.sum((xx <= log10_fdr_threshold) & (yy <= log10_fdr_threshold))\n",
    "    \n",
    "    xx = np.maximum(0, xx + x_jitter_scale * np.max(xx) * rng.randn(len(xx)))\n",
    "    yy = np.maximum(0, yy + y_jitter_scale * np.max(yy) * rng.randn(len(yy)))\n",
    "\n",
    "    scatter = ax.scatter(\n",
    "        xx,\n",
    "        yy,\n",
    "        color='gray',\n",
    "        # c=log10_tpm,\n",
    "        cmap=plt.cm.Reds,\n",
    "        s=2, alpha=0.5)\n",
    "\n",
    "    ax.axhline(log10_fdr_threshold, lw=1, linestyle='--')\n",
    "    ax.axvline(log10_fdr_threshold, lw=1, linestyle='--')\n",
    "\n",
    "    for manifest, highlighted_gene_names in highlight_map.items():\n",
    "        label = manifest[0]\n",
    "        color = manifest[1]\n",
    "        indices = [\n",
    "            gene_names_list.index(gene_name)\n",
    "            for gene_name in highlighted_gene_names]\n",
    "        ax.scatter(\n",
    "            xx[indices],\n",
    "            yy[indices],\n",
    "            s=20,\n",
    "            color=color,\n",
    "            label=label,\n",
    "            marker='o',\n",
    "            facecolor=[1, 1, 1, 0],\n",
    "            linewidths=1)\n",
    "\n",
    "    ax.set_xlabel(r'$-log_{10}~P_{DE}$')\n",
    "    ax.set_ylabel(r'$-log_{10}~P_{DS}$')\n",
    "    \n",
    "    ax.text(\n",
    "        0.5 * log10_fdr_threshold, 0.5 * (np.max(yy) + log10_fdr_threshold),\n",
    "        str(n_ds),\n",
    "        fontsize=8,\n",
    "        rotation='horizontal',\n",
    "        ha='center',\n",
    "        color='red',\n",
    "        bbox=dict(boxstyle='round', facecolor='white', alpha=0.5))\n",
    "\n",
    "    ax.text(\n",
    "        0.5 * log10_fdr_threshold, 0.5 * log10_fdr_threshold,\n",
    "        str(n_boring),\n",
    "        fontsize=8,\n",
    "        rotation='horizontal',\n",
    "        ha='center',\n",
    "        color='red',\n",
    "        bbox=dict(boxstyle='round', facecolor='white', alpha=0.5))\n",
    "\n",
    "    ax.text(\n",
    "        0.5 * (np.max(xx) + log10_fdr_threshold), 0.5 * log10_fdr_threshold,\n",
    "        str(n_de),\n",
    "        fontsize=8,\n",
    "        rotation='horizontal',\n",
    "        ha='center',\n",
    "        color='red',\n",
    "        bbox=dict(boxstyle='round', facecolor='white', alpha=0.5))\n",
    "\n",
    "    ax.text(\n",
    "        0.5 * (np.max(xx) + log10_fdr_threshold), 0.5 * (np.max(yy) + log10_fdr_threshold),\n",
    "        str(n_de_ds),\n",
    "        fontsize=8,\n",
    "        rotation='horizontal',\n",
    "        ha='center',\n",
    "        color='red',\n",
    "        bbox=dict(boxstyle='round', facecolor='white', alpha=0.5))\n",
    "\n",
    "    ax.set_title(group_ids[group_idx] + \" vs. rest\")\n",
    "\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.savefig(\n",
    "    os.path.join(\n",
    "        repo_root, fig_output_root, output_prefix_full + \"_per_cluster_ds_de.pdf\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "pytorch-gpu.1-7.m65",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/pytorch-gpu.1-7:m65"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
